### PR TITLE
CORE - 1374 Allow users to copy links and paths

### PR DESCRIPTION
**Pull request by sri2k1us**
- Allow users to copy Data URL, path from row dot menu in listing and detailed search results.
- Allow users to copy Apps URL from row dot menu from listing and detailed search results.
- Data Search results:
image
- App Search results:
image

- Data View:
image

- Apps View:
image

### CODE REVIEW

File: src/components/apps/listing/Listing.js [hunk 1]
```diff
@@ -196,8 +196,13 @@ function Listing(props) {
...
     useEffect(() => {
-        setAddToBagEnabled(selected && selected.length > 0);
-    }, [selected]);
+        const selApps = getSelectedApps();
+        setAddToBagEnabled(
+            selApps &&
+                selected.length > 0 &&
+                selApps?.filter((app) => app.is_public === true).length === 0
+        );
+    }, [getSelectedApps, selected, selectedApp]);
```
Comment (sri2k1us) [inline src/components/apps/listing/Listing.js:205]: Enable `Add to Bag` only for private apps.

File: src/components/apps/details/Drawer.js [hunk 2]
```diff
@@ -98,33 +99,31 @@ function DetailsHeader({
...
-                    </Tooltip>
-                </div>
-            )}
+                )}
+                <Tooltip title={t("linkToThisApp", { name: appName })}>
+                    <IconButton size="small" onClick={() => setOpen(true)}>
+                        <LinkIcon color="primary" fontSize="small" />
+                    </IconButton>
+                </Tooltip>
+            </div>
```
Comment (sri2k1us) [inline src/components/apps/details/Drawer.js:126]: Allow users to copy link for all apps.

File: src/components/apps/listing/Listing.js [hunk 3]
```diff
@@ -196,8 +196,13 @@ function Listing(props) {
     }, [selected]);
 
     useEffect(() => {
-        setAddToBagEnabled(selected && selected.length > 0);
-    }, [selected]);
+        const selApps = getSelectedApps();
+        setAddToBagEnabled(
+            selApps &&
+                selected.length > 0 &&
+                selApps?.filter((app) => app.is_public === true).length === 0
```
Comment (psarando) [inline src/components/apps/listing/Listing.js:203]: No need to explicitly check `=== true`:
```suggestion
                selApps?.filter((app) => app.is_public).length === 0
```
Or slightly shorter:
```suggestion
                !selApps?.find((app) => app.is_public)
```

File: src/components/data/listing/RowDotMenu.js [hunk 4]
```diff
@@ -111,6 +116,54 @@ function RowDotMenu(props) {
...
+                        baseId={baseId}
+                        onClose={onClose}
+                        onCopyLinkSelected={() => {
+                            const link = `${getHost()}/${
+                                NavigationConstants.DATA
+                            }/${constants.DATA_STORE_STORAGE_ID}${
+                                resource?.path
+                            }?type=${resource?.type}&resourceId=${
+                                resource?.id
+                            }`;
```
Comment (psarando) [inline src/components/data/listing/RowDotMenu.js:131]: This is a bit hard to parse. Maybe it could be refactored into a data util function.

In fact, this string building is repeated at least once, in the data search results changes included in this PR.

File: src/components/data/listing/RowDotMenu.js [hunk 5]
```diff
@@ -111,6 +116,54 @@ function RowDotMenu(props) {
...
+                    />,
+                    <CopyPathMenuItem
+                        key={build(baseId, ids.COPY_PATH_MENU_ITEM)}
+                        baseId={baseId}
+                        onClose={onClose}
+                        onCopyPathSelected={() => {
+                            const copyPromise = copyStringToClipboard(
+                                resource?.path
+                            );
+                            copyPromise.then(
```
Comment (psarando) [inline src/components/data/listing/RowDotMenu.js:144]:
```suggestion
                            copyStringToClipboard(resource?.path).then(
```

File: src/components/utils/CopyLinkButton.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ * @author sriram
+ *
+ * A common CopyLink menu item to use across Data and Apps
+ *
+ */
+import React from "react";
+import { useTranslation } from "i18n";
+import { build } from "@cyverse-de/ui-lib";
+import { IconButton, Tooltip } from "@material-ui/core";
+import { FileCopy } from "@material-ui/icons";
```
Comment (psarando) [inline src/components/utils/CopyLinkButton.js:11]: I'm used to seeing a clipboard icon in places like GitHub for copy-to-clipboard actions, but the only MUI icon that looks like a clipboard is the `Assignment` icon 🤔 

This `FileCopy` icon looks good in this PR's screenshots, though 👍

File: src/components/apps/details/Drawer.js [hunk 7]
```diff
@@ -98,33 +99,31 @@ function DetailsHeader({
     const [open, setOpen] = useState(false);
 
     useEffect(() => {
-        const protocol = window.location.protocol.concat("//");
-        const host = protocol.concat(window.location.host);
-        setLink(
-            host.concat(`/${NavigationConstants.APPS}/${systemId}/${appId}`)
-        );
+        const host = getHost();
+        setLink(`${host}/${NavigationConstants.APPS}/${systemId}/${appId}`);
```
Comment (psarando) [inline src/components/apps/details/Drawer.js:103]: Since this formatted string building is repeated 3 times in this PR, it should probably be refactored into a util function.

**Review by psarando — APPROVED** [review]
LGTM 👍 

I just had a couple of minor refactoring suggestions.

File: src/components/data/menuItems/CopyPathMenuItem.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+import ids from "../ids";
+
+export default function CopyPathMenuItem(props) {
+    const { baseId, onClose, onCopyPathSelected } = props;
+    const { t } = useTranslation("data");
+
+    return (
+        <MenuItem
+            key={build(baseId, ids.COPY_MD5_MENU_ITEM)}
+            id={build(baseId, ids.COPY_MD5_MENU_ITEM)}
```
Comment (ash3rz) [inline src/components/data/menuItems/CopyPathMenuItem.js:21]: 🤔 md5 menu item?  I saw a `COPY_PATH_MENU_ITEM` id was added, seems like that was intended to be used here

File: src/components/search/detailed/AppSearchResults.js [hunk 9]
```diff
@@ -117,14 +122,37 @@ export default function AppSearchResults(props) {
...
+                                    onClick={() => setDetailsApp(original)}
+                                    size="small"
+                                    color="primary"
+                                >
+                                    <Info fontSize="small" />
+                                </IconButton>
+                            </Grid>
+                            <Grid item>
+                                <CopyLinkButton
+                                    onCopyLinkSelected={() => {
```
Comment (ash3rz) [inline src/components/search/detailed/AppSearchResults.js:143]: Minor, but both of these buttons could use IDs for testing

File: src/components/search/detailed/DataSearchResults.js [hunk 10]
```diff
@@ -167,20 +193,73 @@ function DataSearchResults(props) {
...
+                                        copyLinkToClipboardHandler(
+                                            i18nCommon,
+                                            copyPromise
+                                        );
+                                    }}
+                                />
+                            </Grid>
+                            <Grid item>
+                                <CopyPathButton
+                                    onCopyPathSelected={() => {
```
Comment (ash3rz) [inline src/components/search/detailed/DataSearchResults.js:229]: Minor, but these 3 buttons could use IDs

**Review by ash3rz — APPROVED** [review]
LGTM :shipit:

File: src/components/search/detailed/AppSearchResults.js (hunk 9 above)
Comment (sri2k1us) [reply src/components/search/detailed/AppSearchResults.js:143]: Has an id defined https://github.com/sri2k1us/sonora/blob/8e41c8d307fb7087b243c234181216e38235175f/src/components/utils/CopyLinkButton.js#L21

**Comment by sri2k1us** [comment]
Thanks for the reviews.
